### PR TITLE
untie `getColumnKey` function from code not usable in static viz rendering environment

### PR DESCRIPTION
**Pull request by alxnddr**
## Changes

1) Rename `keyForColumn` to `getColumnKey`
2) Split and extract it from the code that is cluttered with dependencies that are impossible to use in the static viz rendering environment

Each series settings is stored in `visualization_settings.series_settings[seriesKey]` so in the static viz environment we should be able to retrieve the same series key. The problem is that the `metabase-lib` code is highly coupled which does not allow importing this single function in the static viz code.

### How to verify

Everything should work as before, specs should not fail

### CODE REVIEW

File: frontend/src/metabase-types/api/mbql/references.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
+export type ReferenceOptionsKeys =
+  | "source-field"
+  | "base-type"
+  | "join-alias"
+  | "temporal-unit"
+  | "binning";
+
+export type ReferenceOptions = Partial<
```
Comment (ranquild) [inline frontend/src/metabase-types/api/mbql/references.ts:8]: With this approach it won't be possible to type options correctly. I believe we should use a regular type for it, e.g. 

```
interface ReferenceOptions {
  binning?: BinningOptions;
  "temporal-unit"?: TemporalUnit;
}

export type TemporalUnit =
  | "default"
  | "minute"
  | "minute-of-hour"
  | "hour"
  | "hour-of-day"
  | "day"
  | "day-of-week"
  | "day-of-month"
  | "day-of-year"
  | "week"
  | "week-of-year"
  | "month"
  | "month-of-year"
  | "quarter"
  | "quarter-of-year"
  | "year";


type BinningOptions = 
  | DefaultBinningOptions
  | NumBinsBinningOptions
  | BinWidthBinningOptions;

interface DefaultBinningOptions {
  strategy: "default";
}

interface NumBinsBinningOptions {
  strategy: "num-bins";
  "num-bins": number;
}

interface BinWidthBinningOptions {
   strategy: "bin-width"; 
  "bin-width": number;
}
```

search in `metabase-types/types` for other types

File: frontend/src/metabase-types/api/dataset.ts [hunk 2]
```diff
@@ -6,11 +6,14 @@ export type RowValue = string | number | null | boolean;
 export type RowValues = RowValue[];
 
 export interface DatasetColumn {
+  id?: number;
   display_name: string;
   source: string;
   name: string;
   remapped_to_column?: DatasetColumn;
   unit?: DatetimeUnit;
+  field_ref?: any;
```
Comment (ranquild) [inline frontend/src/metabase-types/api/dataset.ts:15]: Why it's `any` here?

File: frontend/src/metabase-lib/lib/queries/utils/get-column-key.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  getDimensionReferenceWithoutOptions,
+  getNormalizedDimensionReference,
+  hasStringFieldName,
+  isAggregationReference,
+  isExpressionReference,
+  isFieldReference,
+  isValidDimensionReference,
+} from "metabase-lib/lib/references";
+
+export const getColumnKey = (
```
Comment (ranquild) [inline frontend/src/metabase-lib/lib/queries/utils/get-column-key.ts:14]: This is meant to get the column key for the `visualization_settings.columns`, right? I don't get it how this function could work without handling duplicate column names (which would require the entire query). Is this done somewhere else?

Comment (ranquild) [reply frontend/src/metabase-lib/lib/queries/utils/get-column-key.ts:14]: Currently this logic is in `StructuredQuery` https://github.com/metabase/metabase/blob/25c3050f30cca0309a011e39d15d4350889f05cf/frontend/src/metabase-lib/lib/queries/StructuredQuery.ts#L1405

**Review by ranquild — CHANGES_REQUESTED** [review]
Requesting changes mostly to get answers for the questions in comments

File: frontend/src/metabase-lib/lib/queries/utils/get-column-key.ts (hunk 3 above)
Comment (alxnddr) [reply frontend/src/metabase-lib/lib/queries/utils/get-column-key.ts:14]: It works with columns coming from the BE which have valid column names:
Screen Shot 2022-10-17 at 1 53 44 PM

File: frontend/src/metabase-types/api/dataset.ts (hunk 2 above)
Comment (alxnddr) [reply frontend/src/metabase-types/api/dataset.ts]: fixed

File: frontend/src/metabase-types/api/mbql/references.ts (hunk 1 above)
Comment (alxnddr) [reply frontend/src/metabase-types/api/mbql/references.ts]: fixed

File: frontend/src/metabase-types/api/mbql/references.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
+export type ReferenceOptionsKeys =
```
[the remark this answers is not in the document]
Comment (alxnddr) [reply frontend/src/metabase-types/api/mbql/references.ts]: moved the types

**Review by ranquild — APPROVED** [review]

**Pull request merged**